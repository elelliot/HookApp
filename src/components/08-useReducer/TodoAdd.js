import React from 'react'
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ handleAddTodo }) => {
    
    const [ { description }, handleInputChange, reset ] = useForm({
        description:'',
    });


    //Agregamos un todo
    const handleSubmit = (e) => {
        e.preventDefault();

        if( description.trim().length <= 1 ) {
            return;
        }

        const newTodo = {
            id: new Date().getTime(),
            desc: description,
            done: false 
        };

        handleAddTodo( newTodo );

        console.log(`Se agrego la actividad ${ newTodo.desc }\nCon id: ${ newTodo.id } `);
        reset();
    }

    return (
        <>
            <h4>Agregar TODO</h4>
            <hr />

            <form onSubmit={ handleSubmit }>

                <input
                    type="text"
                    name="description"
                    className="form-control"
                    placeholder="Aprender..."
                    autoComplete="off"
                    value={ description }
                    onChange = { handleInputChange }
                />
                
                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>

            </form>
        </>
    )
}
