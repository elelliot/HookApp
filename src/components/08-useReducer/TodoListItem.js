import React from 'react'

export const TodoListItem = ({ todo, i, handleDelete, handleToggle }) => {
    return (
        <div>
            <li
                key = { todo.id }
                className= "list-group-item"
            > 
                <p 
                    className = { `${ todo.done && 'complete' }` /*si el todo.done es true, regresa la clase complete*/ }
                    onClick= {() => handleToggle( todo.id ) }
                > { i + 1 }. { todo.desc } </p>
                <button
                    className="btn btn-danger"
                    onClick={ () => handleDelete( todo.id, todo.desc ) }
                >
                    Borrar
                </button>
            </li>
        </div>
    )
}
