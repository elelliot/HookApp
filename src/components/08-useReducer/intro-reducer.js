
const initialState = [{
    id: 1,
    todo: 'Comprar Pan',
    done: false
}];

const todoReducer = ( state = initialState, action ) => {
    
    if( action?.type === 'agregar' ){ //si hay un tipo, leelo, si no no hagas nada (b√°sicamente evitamos que crashee)
        console.log( action.type );
        return [...state, action.payload]; //no usemos el push en react por que mutamos el objeto
    }
       
    return state;
}


let todos = todoReducer();


const newTodo = {
    id: 2,
    todo: 'Comprar Leche',
    done: false
}

const agregarTodoAction = {
    type: 'agregar',
    payload: newTodo

}


todos = todoReducer( todos, agregarTodoAction );


console.log(todos);