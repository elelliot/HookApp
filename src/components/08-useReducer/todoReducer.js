

export const todoReducer = ( state = [] , action ) => {

    switch ( action.type ) {
        case 'add':
            return [...state , action.payload ];

        case 'delete':
            return state.filter( todo => todo.id !== action.payload );

        case 'toggle':
            return state.map( todo => 
                // forma con operador ternario, debe regresar 2 valores
                // extraemos el valor del todo que coincida con la condicion y cambiamos el done
                // y en el caso que sean iguales entonces no hago nada
                ( todo.id === action.payload )
                    ? {...todo, done: !todo.done }
                    : todo    
            )    


        // forma larga
        // return state.map( todo => {

            //     if( todo.id === action.payload ){
            //         return {
            //             ...todo,
            //             done: !todo.done
            //         }
            //     } else {
            //         return todo;
            //     }
            // })

        default:
            return state;
    }
}