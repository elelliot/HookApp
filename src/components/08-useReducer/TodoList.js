import React from 'react'
import PropTypes from 'prop-types';
import { TodoListItem } from './TodoListItem'

export const TodoList = ({ todos, handleDelete, handleToggle }) => {
    return (
        <div>
            <ul className="list-group list-group-flush">
                {
                    todos.map( (todo, i) => (
                        
                        <TodoListItem 
                            key = { todo.id }
                            todo = { todo }
                            i = { i }
                            handleDelete= { handleDelete } 
                            handleToggle= { handleToggle } 
                        />
                    ))
                }
            </ul>
        </div>
    )
}

TodoList.propTypes = {
    todos: PropTypes.array.isRequired,
    handleDelete: PropTypes.func.isRequired,
    handleToggle: PropTypes.func.isRequired
}
