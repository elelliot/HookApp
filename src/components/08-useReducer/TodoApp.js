import React, { useEffect, useReducer } from 'react'
import { todoReducer } from './todoReducer';

import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';
import './styles.css';

// const initialState = [{
//     id: new Date().getTime(),
//     desc: 'Aprender React',
//     done: false
// }];



// Si se da refresh en el navegador, la funcion init lee el local storage y establece el estado inicial del reducer

// cualquier cosa que retorne la funcion init, es el initialState
const init= () => {
    
    //localStorage.getItem regresa la información null si un objeto no existe
    //JSON.parse de algo que es null, regresa null, si no , lo regresa formateado

    //Entonces si existen todos regresalos como arreglo y si es null, regresar un arreglo vacío que será el estado inicial

    return JSON.parse(localStorage.getItem('todos')) || [];//

}

export const TodoApp = () => {


    //dispatch es una funcion que le mandan una accion
    const [ todos, dispatch ] = useReducer(todoReducer, [], init);

    
    //El efecto graba en el localStorage cualquier cambio que suceda en los todos, agregar edite elimine
    useEffect( () => {
        
        //Solo se pueden grabar Strings en el local storage, por tanto lo formateamos a JSON el objeto
        // Se guarda en un formato Key, Value
        localStorage.setItem('todos', JSON.stringify( todos ) )
    }, [todos] );

    // console.log( description );


    const handleDelete = ( todoId, todoDesc ) => {
        console.log(`Actividad: ${todoDesc}\nId: ${todoId}\nELIMINADO`);

        const action = {
            type: 'delete',
            payload: todoId
        }
        
        dispatch( action );
    }


    const handleToggle = ( todoId ) => {

        dispatch( {
            type: 'toggle',
            payload: todoId
        });
    }

    const handleAddTodo = ( newTodo ) => {

        dispatch( {
            type: 'add',
            payload: newTodo
        });

    }

    return (
        <div>
            <h1>TodoApp ( { todos.length } )</h1>
            <hr />

            <div className= "row">

                <div className="col-7">

                    <TodoList 
                        todos = { todos } 
                        handleDelete = { handleDelete } 
                        handleToggle= { handleToggle } 
                    />

                </div>


                <div className="col-5">
                    
                    <TodoAdd 
                        handleAddTodo = { handleAddTodo } 
                    />

                </div>

            </div>
        </div>
    )
}
