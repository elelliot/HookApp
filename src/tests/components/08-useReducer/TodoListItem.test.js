import { shallow } from "enzyme"
import { TodoListItem } from "../../../components/08-useReducer/TodoListItem";
import { demoTodos } from "../../fixtures/demoTodos";


describe('Pruebas en <TodoListItem />', () => {
    //solo para saber si son llamadas
    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow( 
        <TodoListItem
            todo = { demoTodos[0] }
            i ={ 0 }
            handleDelete = { handleDelete } 
            handleToggle = { handleToggle }
        /> 
        );


    test('Debe mostrarse correctamente', () => {
        expect( wrapper ).toMatchSnapshot();
    });


    test('Debe llamar la función handleDelete', () => {
        //Para que el handleDelete sea llamado, se debe dar click al botón delete
        wrapper.find('button').simulate('click');

        //handleDelete debe ser llamado
        //expect( handleDelete ).toHaveBeenCalledWith( demoTodos[0].id);
        expect( handleDelete ).toHaveBeenCalledWith( demoTodos[0].id, demoTodos[0].desc );

    });
    

    test('Debe llamar la función handleToggle', () => {
        wrapper.find('p').simulate('click');
        expect( handleToggle ).toHaveBeenCalledWith( demoTodos[0].id );

    });

    test('Debe mostrar el texto correctamente del <p>', async () => {
        //contenido del parrafo ("indice"+"."+ " todo")
        const p = wrapper.find('p');
        expect( p.text().trim() ).toBe(`1. ${ demoTodos[0].desc }`);
    });

    test('Debe tener class= complete si el todo.done= TRUE ', async () => {
        //Asignamos un toDo por defecto y le damos el valor true
        const todo= demoTodos[0];
        todo.done= true;
    

        //solo manejaremos el todo
        const wrapper = shallow( 
            <TodoListItem
                todo = { todo }
            /> 
        );
        //Para ver que el class = complete
        console.log( wrapper.html() );
        expect( wrapper.find('p').hasClass('complete') ).toBe( true );//o toBeTruthy
    });
    
    
})
