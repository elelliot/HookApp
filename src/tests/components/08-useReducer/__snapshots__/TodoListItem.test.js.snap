// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoListItem /> Debe mostrarse correctamente 1`] = `
<div>
  <li
    className="list-group-item"
    key="1"
  >
    <p
      className="false"
      onClick={[Function]}
    >
       
      1
      . 
      JavaScript
       
    </p>
    <button
      className="btn btn-danger"
      onClick={[Function]}
    >
      Borrar
    </button>
  </li>
</div>
`;
